Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1833,78 +1835,153 @@ inline void SetImage(CGImageRef image) {
                                    size_t height,
                                    size_t bitsPerComponent,
                                    size_t bytesPerRow,
                                    CGColorSpaceRef colorSpace,
                                    CGBitmapInfo bitmapInfo) {
-    UNIMPLEMENTED();
-    return nullptr;
+    return CGBitmapContextCreateWithData(data, width, height, bitsPerComponent, bytesPerRow, colorSpace, bitmapInfo, nullptr, nullptr);
+}
+
+/**
+ @Status Caveat
+ @Notes releaseCallback and releaseInfo is ignored. Also if the image colorspace is not a valid render target format (as per
+ https://msdn.microsoft.com/en-us/library/windows/desktop/dd756766(v=vs.85).aspx#supported_wic_formats), it's converted to 32bpp BGRA
+ format. Does not support conversion if data is already provided.
+*/
+CGContextRef CGBitmapContextCreateWithData(void* data,
+                                           size_t width,
+                                           size_t height,
+                                           size_t bitsPerComponent,
+                                           size_t bytesPerRow,
+                                           CGColorSpaceRef space,
+                                           uint32_t bitmapInfo,
+                                           CGBitmapContextReleaseDataCallback releaseCallback,
+                                           void* releaseInfo) {
+    RETURN_NULL_IF(!width);
+    RETURN_NULL_IF(!height);
+    RETURN_NULL_IF(!space);
+
+    size_t bitsPerPixel = ((bytesPerRow / width) << 3);
+    REFGUID pixelFormat = _CGImageGetWICPixelFormatFromImageProperties(bitsPerComponent, bitsPerPixel, space, bitmapInfo);
+
+    ComPtr<IWICBitmap> customBitmap;
+
+    if (_CGIsValidRenderTargetPixelFormat(pixelFormat)) {
+        // if data is null, enough memory is allocated via CGIWICBitmap
+        customBitmap = Make<CGIWICBitmap>(data, pixelFormat, height, width);
+    } else {
+        // TODO #<GITHUB-ID>: this will be an issue if we have change in stride, account for that.
+        // Also as per documentation, it's best to leave CGBitmapContext to manage the memory
+        if (data != nullptr) {
+            return nullptr;
+        }
+        customBitmap = Make<CGIWICBitmap>(data, GUID_WICPixelFormat32bppPBGRA, height, width);
+    }
+
+    RETURN_NULL_IF(!customBitmap);
+
+    woc::unique_cf<CGImageRef> image(_CGImageCreateWithWICBitmap(customBitmap.Get()));
+    RETURN_NULL_IF(!image);
+
+    ComPtr<ID2D1Factory> factory;
+    RETURN_NULL_IF_FAILED(_CGGetD2DFactory(&factory));
+
+    ComPtr<ID2D1RenderTarget> renderTarget;
+    RETURN_NULL_IF_FAILED(factory->CreateWicBitmapRenderTarget(customBitmap.Get(), D2D1::RenderTargetProperties(), &renderTarget));
+    return _CGBitmapContextCreateWithRenderTarget(renderTarget.Get(), image.get());
+}
+
+/**
+ @Status Interoperable
+*/
+CGBitmapInfo CGBitmapContextGetBitmapInfo(CGContextRef context) {
+    NOISY_RETURN_IF_NULL(context, 0);
```

Comment:
Is 0 the default bitmap info? Should this be a constant? And do all of these match reference platform?